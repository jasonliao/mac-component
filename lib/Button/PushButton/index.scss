.push-button {
  box-sizing: border-box;
  color: rgb(38, 38, 38);
  border: 1px solid rgb(200, 200, 200);
  background: #fff;
  border-radius: 4px;
  outline: none;
}

.push-button:active {
  background: linear-gradient(to bottom, rgb(75, 145, 255), 55%, rgb(2, 92, 228));
  color: rgb(222, 234, 253);
  border: 1px solid rgb(69, 142, 251);
}

.push-button__on {
  background: linear-gradient(to bottom, rgb(107, 165, 261), 55%, rgb(4, 107, 255));
  border: 1px solid rgb(107, 165, 261);
  color: #fff;
  font-weight: 500
}

// TODO: use ::after to add gradient border `.push-button__on` and `.push-button:action`
// https://stackoverflow.com/questions/5706963/possible-to-use-border-radius-together-with-a-border-image-which-has-a-gradient

.push-button__disabled {
  background: rgb(246, 246, 246);
  border: 1px solid rgb(213, 213, 213);
  color: rgb(190, 190, 190);
}

.push-button.push-button__disabled:active {
  background: rgb(246, 246, 246);
  border: 1px solid rgb(213, 213, 213);
  color: rgb(190, 190, 190);
}

.push-button__regular {
  padding: 1px 14px;
  font-size: 13px;
}

.push-button__small {
  padding: 1px 14px;
  font-size: 11px;
}

.push-button__mini {
  padding: 1px 14px;
  font-size: 9px;
}