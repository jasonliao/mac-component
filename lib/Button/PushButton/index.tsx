import * as React from 'react'
import classNames from 'classnames'
import './index.scss'

interface Props {
  text?: string
  size?: string
  type?: string
  disabled?: boolean
  style?: object
  className?: string
  onClick?: (e: React.MouseEvent) => void
}

export default class PushButton extends React.Component<Props, any> {
  static defaultProps = {
    text: '',
    size: 'regular',
    type: 'off',
    disabled: false,
    style: {},
    className: '',
    onClick: () => {}
  }

  handleClick = (e: React.MouseEvent) => {
    const { onClick, disabled } = this.props
    !disabled && onClick(e)
  } 

  render() {
    const { text, size, disabled, type, style, className } = this.props
    return (
      <button
        onClick={this.handleClick}
        className={classNames(
          className,
          'push-button',
          `push-button__${size}`,
          { 
            'push-button__disabled': disabled,
            'push-button__on': type === 'on' && !disabled
          }
        )}
        style={{ ...style }}
      >
        { text }
      </button>
    )
  }
}
