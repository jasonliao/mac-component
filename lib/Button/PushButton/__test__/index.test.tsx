import * as React from 'react'
import { shallow } from 'enzyme'
import * as renderer from 'react-test-renderer'

import PushButton from '../index'

describe('PushButton Component', () => {
  it('render text', () => {
    const props = { text: 'PushButton' }
    const wrapper = shallow(<PushButton { ...props } />)

    const text = wrapper.find('button').text()
    expect(text).toBe(props.text)
  })

  it('render with mini size', () => {
    const props = { size: 'mini' }
    const wrapper = shallow(<PushButton { ...props } />)
    const hasClass = wrapper.find('.push-button').hasClass(`push-button__${props.size}`)

    expect(hasClass).toBe(true)
  })

  it('render with type', () => {
    const props = { type: 'on' }

    const wrapper = shallow(<PushButton { ...props } />)
    const hasClass = wrapper.find('.push-button').hasClass(`push-button__${props.type}`)

    expect(hasClass).toBe(true)
  })

  it('render disabled button', () => {
    const props = { disabled: true }

    const wrapper = shallow(<PushButton { ...props } />)
    const hasClass = wrapper.find('.push-button').hasClass(`push-button__disabled`)

    wrapper.simulate('click')

    expect(hasClass).toBe(true)
  })

  it('button clicked', () => {
    const props = { onClick: jest.fn() }
    const wrapper = shallow(<PushButton { ...props } />)
    wrapper.simulate('click')

    expect(props.onClick).toHaveBeenCalledTimes(1)
  })

  it('disabled button clicked', () => {
    const props = { disabled: true, onClick: jest.fn() }

    const wrapper = shallow(<PushButton { ...props } />)
    wrapper.simulate('click')

    expect(props.onClick).toHaveBeenCalledTimes(0)
  })

  it('match the snapshot', () => {
    const props = {
      text: 'PushButton',
      type: 'on',
      size: 'small',
      style: { width: '100px' },
      className: 'custom_classname'
    }
    const tree = renderer.create(<PushButton { ...props } />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})