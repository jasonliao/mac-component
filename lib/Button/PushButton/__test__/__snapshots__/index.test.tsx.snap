// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PushButton Component match the snapshot 1`] = `
<button
  className="custom_classname push-button push-button__small push-button__on"
  onClick={[Function]}
  style={
    Object {
      "width": "100px",
    }
  }
>
  PushButton
</button>
`;
